### ISSUE TITLE
Thread#current and Fiber#transfer gets the wrong Ruby Thread due to Fiber pooling

### ISSUE THREAD

**Issue by eregon** [graalvm/truffleruby#903]
Originally discovered by @chuckremes with the script:
https://gist.github.com/chuckremes/14c9ef06efc0bc62b9dc5b588c06719f

A similar bug seems to also happen on Rubinius:
https://github.com/rubinius/rubinius/issues/3776

**Comment by eregon** [comment]
Also on JRuby, although it fails differently: https://github.com/jruby/jruby/issues/4920

**Comment by eregon** [comment]
Buggy output on TruffleRuby for the record:
```
Step 1
Step 2
Step 3
Step 4
Step 5
16
Step 6
Step 7
Step 8
2
Starting steps over from new thread, using new Foo object and fibers
Step 1
Step 2
Step 3
Step 4
g.rb:9:in `transfer': fiber called across threads (FiberError)
	from g.rb:9:in `block in initialize'
```

**Comment by eregon** [comment]
Changing https://github.com/graalvm/truffleruby/blob/b005dd4ea04b94180ed66977c121f38f6730efee/src/main/java/org/truffleruby/core/thread/GetCurrentRubyThreadNode.java#L29 to `"0"` or removing the specialization fixes it.
But I'm looking if we can do something better.

**Comment by eregon** [comment]
Specs and fix in #904.
